src/nile/core/types/transaction.py

```diff
@@ -0,0 +1,343 @@
+    network: str = "localhost"
+    version: int = TRANSACTION_VERSION
+    overriding_path: str = None
+
+    # Public fields not expected in construction time
+    hash: int = field(init=False, default=0)
+    query_hash: int = field(init=False, default=0)
+    chain_id: int = field(init=False)
+    contract_class: str = field(init=False, default=None)
+    entry_point_selector: int = field(init=False)
+
+    # Internal fields
+    _udc_deployment_address: int = field(init=False, default=None)
+
+    def __post_init__(self):
+        """Initialize hash, entry_point_selector, contract_class and chain_id."""
+        self.chain_id = (
+            StarknetChainId.MAINNET.value
+            if self.network == "mainnet"
+            else StarknetChainId.TESTNET.value
+        )
+
+        if self.tx_type == "invoke":
+            get_tx_hash_func = self._get_invoke_hash
+        elif self.tx_type == "declare":
+            self.contract_class = get_contract_class(
+                contract_name=self.contract_to_submit,
+                overriding_path=self.overriding_path,
+            )
+            get_tx_hash_func = self._get_declare_hash
+        elif self.tx_type == "deploy_account":
+            get_tx_hash_func = self._get_deploy_account_hash
+
+        self.hash = get_tx_hash_func()
+        self.query_hash = get_tx_hash_func(QUERY_VERSION_BASE + self.version)
+        self.entry_point_selector = get_selector_from_name(self.entry_point)
+
+        # Validate the transaction object
+        self._validate()
+
+    @classmethod
+    def create_invoke(
+        cls,
+        account_address: int,
+        calldata,
+        max_fee: int,
+        nonce: int,
+        network: int,
+        version: int = TRANSACTION_VERSION,
+    ) -> "Transaction":
+        """Create an invoke transaction."""
+        return cls(
+            tx_type="invoke",
+            account_address=account_address,
+            calldata=calldata,
+            max_fee=max_fee,
+            nonce=nonce,
+            network=network,
+            version=version,
+        )
+
+    @classmethod
+    def create_declare(
+        cls,
+        account_address: int,
+        contract_to_submit: str,
+        max_fee: int,
+        nonce: int,
+        network: int,
+        version: int = TRANSACTION_VERSION,
+    ) -> "Transaction":
+        """Create a declare transaction."""
+        return cls(
+            tx_type="declare",
+            account_address=account_address,
+            contract_to_submit=contract_to_submit,
+            max_fee=max_fee,
+            nonce=nonce,
+            network=network,
+            version=version,
+        )
+
+    @classmethod
+    async def create_udc_deploy(
+        cls,
+        account,
+        contract_name: str,
+        salt: int,
+        unique: bool,
+        calldata,
+        deployer_address: int,
+        max_fee: int,
+        nonce: int = None,
+        overriding_path=None,
+    ) -> "Transaction":
```

Comment:
i'm not sure the UDC needs this special treatment. i'd leave the Transaction types as similar as the "real transactions" as possible so it's easier to track changes at the protocol level. UDC deployments are just a special case of invoke transactions.

the question is where should the UDC abstraction be implemented? i think maybe at the account level, since it's also a solution for accounts. what do you think?

Reply:
> i'm not sure the UDC needs this special treatment. i'd leave the Transaction types as similar as the "real transactions" as possible so it's easier to track changes at the protocol level. UDC deployments are just a special case of invoke transactions.

Agree

> the question is where should the UDC abstraction be implemented? i think maybe at the account level, since it's also a solution for accounts. what do you think?

I feel is probably worth adding another module to handle UDC (creating the invoke `Transaction` form there and providing API for interactions, like getting the predicted deployment address). This should be a fairly small class.